Add CountBy and CountAggregation operations

dlindhol opened the pull request:
The `CountBy` operation regroups the data by the given variables which become the domain of a new dataset. The range is the number of samples with that domain value. This builds on the `GroupByVariable` operation by using an aggregation operation that simply counts the number of samples being aggregated. As a bonus, the `CountAggregation` can be used to return the number of samples in a dataset so I exposed that as a `count` operation.

dlindhol committed:
Add CountBy and CountAggregation operations

dlindhol commented:
Note that "group by" in general can yield multiple elements for each value. **Relational** algebra implementations (e.g. SQL, Spark) often require an aggregation method like sum or average to reduce this set of elements to one. Fortunately, the **functional** data model allows for nested `Function`s. I kept the _aggregation_ abstraction for our `GroupOperation`s (in part because we were working with Spark at the time). It has proven to be quite handy. The default aggregation is to combine the `Sample`s into a nested `Function`. For `CountBy`, I am aggregating that set of `Sample`s into a single `count` variable.

Note, also, that grouping in the functional data model may run into memory problems. We have to process every `Sample` of the original dataset in order to build the new domain set which must be ordered. We'll need to address this some day.

Note, also, also, that `Aggregation` is currently defined in terms of an `Iterable[Sample]`. This aligns with the decision that nested `Function`s are expected to fit in memory. However, we can also use `Aggregation` as a stand-alone operation as I am here for `count`. This currently imposes an unnecessary constraint on the size of a dataset that can be counted. In the interest of "Steams all in", we may want to define the aggregate function in terms of `Stream`.

lindholc commented on core/src/main/scala/latis/ops/CountAggregation.scala:
```diff
@@ -0,0 +1,27 @@
+class CountAggregation() extends Aggregation {
+
+  def aggregateFunction(model: DataType): Iterable[Sample] => Data =
+    (samples: Iterable[Sample]) => Data.LongValue(samples.size.toLong)
```
You don't need the type here, but it doesn't hurt.

I usually leave out types that can be inferred unless it gets confusing.

lindholc commented on core/src/main/scala/latis/ops/CountBy.scala:
```diff
@@ -0,0 +1,26 @@
+import latis.util.Identifier
+import latis.util.LatisException
+
+class CountBy(ids: NonEmptyList[Identifier]) extends GroupByVariable(ids.toList: _*) {
```
Would be good to have a brief description here.

lindholc commented on core/src/main/scala/latis/ops/CountBy.scala:
```diff
@@ -0,0 +1,26 @@
+          .traverse(Identifier.fromString)
+          .map(new CountBy(_)),
+        LatisException("")
+      )
```
By turning things inside out you can keep `id` around if you need to make an error referencing it:

```scala
NonEmptyList(id, ids).traverse { id =>
  Either.fromOption(
    Identifier.fromString(id),
    LatisException(s"Not a valid identifier: $id")
  )
}.map(new CountBy(_))
```

`NonEmptyList.apply` takes a head and a tail, so you don't nee to use the `: _*` syntax.

dlindhol replied:
Oops, looks like I didn't finish that error message.

dlindhol committed:
Clean up and expose Contains operation

lindholc commented on core/src/test/scala/latis/ops/ContainsSuite.scala:
```diff
@@ -0,0 +1,43 @@
+
+import latis.dsl._
+
+class ContainsSuite extends AnyFunSuite {
```
I assume we expect no samples if there are no matches? That might be worth a test.

dlindhol committed:
Drop quotes when interpreting string values

dlindhol committed:
Return Either from Filter predicate construction

This will allow us raise some errors earlier. This is a stopgap
until we make operations safer by construction.

dlindhol commented on core/src/main/scala/latis/ops/Filter.scala:
```diff
@@ -29,27 +30,30 @@ trait Filter extends UnaryOperation with StreamOperation { self =>
    * should be kept.
    */
-  def predicate(model: DataType): Sample => Boolean
+  def predicate(model: DataType): Either[LatisException, Sample => Boolean]
```
I decided to add some tests for the `Contains` operation to see how the failure modes manifested themselves. The answer: not well. I had been hoping that thrown exceptions would magically get raised into Stream errors. I just got stack traces in tests. As a stopgap for my plans for "operations safer by construction" by introducing the model sooner, I chose to modify `Filter` (and thus `Selection` and `Contains`) to return `Either` when constructing the predicate.

dlindhol commented on core/src/main/scala/latis/dsl/package.scala:
```diff
@@ -31,7 +32,7 @@ package object dsl {
     def uncurry(): Dataset                         = dataset.withOperation(Uncurry())
     def curry(n: Int): Dataset                     = dataset.withOperation(Curry(n))
-    def groupByVariable(ids: Identifier*): Dataset = dataset.withOperation(GroupByVariable(ids: _*))
+    def groupByVariable(ids: Identifier*): Dataset = dataset.withOperation(new GroupByVariable(ids: _*))
```
More work to do for consistent operation construction. I'll do this as part of the "safer by construction" plans to include the model so we can vet things earlier.

dlindhol commented on core/src/main/scala/latis/model/ScalarAlgebra.scala:
```diff
@@ -62,9 +63,10 @@ trait ScalarAlgebra { scalar: Scalar =>
-    parseValue(value)
-  //TODO: support units, e.g. "1.2 meters"
+  def convertValue(value: String): Either[LatisException, Datum] = valueType match {
+    case StringValueType => parseValue(StringUtils.removeDoubleQuotes(value))
```
Values for string variables (e.g. in API calls) can now be quoted or unquoted. I only did double quotes here. Not sure if we want to strip single quotes.

dlindhol commented on core/src/main/scala/latis/ops/GroupByVariable.scala:
```diff
@@ -14,7 +14,7 @@ import latis.util.LatisException
  * Does not preserve nested Tuples, for now.
  */
-case class GroupByVariable(ids: Identifier*) extends GroupOperation {
+class GroupByVariable(ids: Identifier*) extends GroupOperation {
```
Can't extend a case class.

dlindhol commented on core/src/main/scala/latis/time/Time.scala:
```diff
@@ -65,7 +65,7 @@ class Time protected (
    * Construct a reusable TimeFormat or UnitConverter for bigger conversion tasks.
    */
-  override def convertValue(value: String): Either[Exception, Datum] = {
+  override def convertValue(value: String): Either[LatisException, Datum] = {
```
Not sure how this slipped by.
